os/include/tinyara/mm/mm.h

```diff
@@ -424,29 +424,36 @@ extern "C" {
  * ARCH_DATA_RESERVE_SIZE
  */
 #include <tinyara/addrenv.h>
-#define BASE_HEAP (&ARCH_DATA_RESERVE->ar_usrheap)
-
-#elif defined(CONFIG_BUILD_PROTECTED) && !defined(__KERNEL__)
-extern uint32_t _stext;
-#define BASE_HEAP ((struct mm_heap_s *)_stext)
+#define USR_HEAP (&ARCH_DATA_RESERVE->ar_usrheap)
 
 #elif defined(CONFIG_BUILD_PROTECTED) && defined(__KERNEL__)
 
 #include <tinyara/userspace.h>
 
 #ifdef CONFIG_APP_BINARY_SEPARATION
+#include <tinyara/sched.h>
 #define USR_HEAP_TCB ((struct mm_heap_s *)((struct tcb_s*)sched_self())->uheap)
-#define USR_HEAP_CFG ((struct mm_heap_s *)(*(uint32_t *)((uint32_t)USERSPACE + sizeof(struct userspace_s))))
-#define BASE_HEAP (USR_HEAP_TCB == NULL ? USR_HEAP_CFG : USR_HEAP_TCB)
+#define USR_HEAP_CFG ((struct mm_heap_s *)(*(uint32_t *)(((struct userspace_s *)USERSPACE)->us_textstart)))
+#define USR_HEAP (USR_HEAP_TCB == NULL ? USR_HEAP_CFG : USR_HEAP_TCB)
+#else
+#define USR_HEAP ((struct mm_heap_s *)(*(uint32_t *)(((struct userspace_s *)USERSPACE)->us_textstart)))
+#endif
+
```

Comment:
Can we change like below to remove duplicate code
#define USR_HEAP_CFG ((struct mm_heap_s *)(*(uint32_t *)(((struct userspace_s *)USERSPACE)->us_textstart)))
#ifdef CONFIG_APP_BINARY_SEPARATION
#include <tinyara/sched.h>
#define USR_HEAP_TCB ((struct mm_heap_s *)((struct tcb_s*)sched_self())->uheap)	
#define USR_HEAP (USR_HEAP_TCB == NULL ? USR_HEAP_CFG : USR_HEAP_TCB)
#else
#define USR_HEAP USR_HEAP_CFG 
#endif

Reply:
I am going to raise a new PR for removing the USERSPACE completely when app separation is enabled. In that case, the above definitions will get much simpler. So, for the time being, let us keep the current code.